src/validators/nested_model.rs

```diff
@@ -0,0 +1,76 @@
+use pyo3::{
+    intern,
+    types::{PyAnyMethods, PyDict, PyDictMethods, PyType},
+    Bound, Py, PyObject, PyResult, Python,
+};
+
+use crate::{definitions::DefinitionsBuilder, errors::ValResult, input::Input};
+
+use super::{BuildValidator, CombinedValidator, SchemaValidator, ValidationState, Validator};
+
+#[derive(Debug, Clone)]
+pub struct NestedModelValidator {
+    model: Py<PyType>,
+    name: String,
+}
+
+impl_py_gc_traverse!(NestedModelValidator { model });
+
+impl BuildValidator for NestedModelValidator {
+    const EXPECTED_TYPE: &'static str = "nested-model";
+
+    fn build(
+        schema: &Bound<'_, PyDict>,
+        _config: Option<&Bound<'_, PyDict>>,
+        _definitions: &mut DefinitionsBuilder<super::CombinedValidator>,
+    ) -> PyResult<super::CombinedValidator> {
+        let py = schema.py();
+        let model = schema
+            .get_item(intern!(py, "model"))?
+            .expect("Invalid core schema for `nested-model` type")
+            .downcast::<PyType>()
+            .expect("Invalid core schema for `nested-model` type")
+            .clone();
+
+        let name = model.getattr(intern!(py, "__name__"))?.extract()?;
+
+        Ok(CombinedValidator::NestedModel(NestedModelValidator {
+            model: model.clone().unbind(),
+            name,
+        }))
+    }
+}
+
+impl Validator for NestedModelValidator {
+    fn validate<'py>(
+        &self,
+        py: Python<'py>,
+        input: &(impl Input<'py> + ?Sized),
+        state: &mut ValidationState<'_, 'py>,
+    ) -> ValResult<PyObject> {
+        self.model
+            .bind(py)
+            .call_method(intern!(py, "model_rebuild"), (), None)
```

Comment:
This ensures that we turn any MockSerializer/Validators into actual ones, there's probably better ways of doing this but I just wanted to make sure it fixed a crash :)

Reply:
Yeah, my hunch here is that we shouldn't use this approach if a model rebuild needs to be done - I don't think we should call `pydantic` specific functions from `pydantic-core`.

Seems like a good pattern to have `pydantic` aware of `pydantic-core`'s implementation details and API, but not the other way around.

Perhaps we could skip using this pattern for this case? Could you give an example of a test case where this needs to happen?

Reply:
This really isn't okay. We can't be calling model rebuild during validation. And yes we should _also_ not be calling pydantic functions from pydantic-core.

IMO the way this should work is that we make a wrapper type that lets you encapsulate a `SchemaValidator` in a Python object.

Reply:
More over, i think we need to store a reference to the inner `SchemaValidator` when building this validator (or worst case, after the first call), otherwise the performance of this will mean we can't use it except in edge cases.

Reply:
We no longer call `model_rebuild` directly from core. Instead the schema contains a callable to get the schema/validator/serializer. This might sometimes involve rebuilding mocks but this seems unavoidable (and I believe is something we already do when trying to validate a model containing forward refs the first time). Hopefully this resolves the concerns around performance of calling rebuild and also the conceptual problems with core depending on pydantic